Endpoints Guide (#1638)
withastro/docs: src/pages/en/core-concepts/endpoints.md

---
layout: ~/layouts/MainLayout.astro

> title: Endpoints

Reviewer: I'm totally fine with calling these Endpoints, but just for reference I've also seen them called **API Routes**. I'm assuming the docs team discussed it and decided on Endpoints, but just in case not, there's another name we've used.

description: Learn how to create endpoints that serve any kind of data
i18nReady: true
---

> Astro lets you create custom endpoints to serve any kind of data. You can use this to generate images, expose an RSS document, or build a full API for your site.

Author: Do you think we should do it here, making "API Routes" a synonym for all custom endpoints, and remove the `(API Routes)` next to `Server Endpoints`? I was thinking that it was specifically the server version that was previously called API Routes, and the static version we called non-HTML pages  (and then file routes).

Or maybe something like `Astro lets you create custom endpoints to serve any kind of data. You can use this to generate images, expose an RSS document, or use them as API Routes to build a full API for your site.` What do you think?

> Astro lets you create custom endpoints to serve any kind of data. You can use this to generate images, expose an RSS document, or build a full API for your site.

Reviewer 2: I like that!

> If your project is in [static](/en/reference/configuration-reference/#output) mode, custom endpoints are called at build time to produce static files. In [SSR](/en/guides/server-side-rendering/) mode, custom endpoints turn into live server endpoints that are called on request. Static and SSR endpoints are defined similarly, but SSR endpoints support additional features.

Author: I like `statically-generated sites`, and I rephrased the second sentence to make it clear that static is the default: 
> `In statically-generated sites, your custom endpoints are called at build time to produce static files. If you opt in to [SSR](/en/guides/server-side-rendering/) mode...`

> If your project is in [static](/en/reference/configuration-reference/#output) mode, custom endpoints are called at build time to produce static files. In [SSR](/en/guides/server-side-rendering/) mode, custom endpoints turn into live server endpoints that are called on request. Static and SSR endpoints are defined similarly, but SSR endpoints support additional features.

Reviewer 2: Yes, I like that!

## Static File Endpoints
To create a custom endpoint, add a `.js` or `.ts` file to the `/pages` directory. The `.js` or `.ts` extension will be removed during the build process, so the name of the file should include the extension of the data you want to create. For example, `src/pages/data.json.ts` will build a `/data.json` endpoint.

> In that file, export a `get` function (optionally `async`) that returns an object with a `body`. In static mode, Astro will call this at build time to generate the corresponding file.
> 
> ```js
> // Example: src/pages/builtwith.json.ts
> // Outputs: /builtwith.json
> 
> // File routes export a get() function, which gets called to generate the file.
> // Return an object with `body` to save the file contents in your final build.
> export async function get({params, request}) {
>   return {
>     body: JSON.stringify({
>       name: 'Astro',
>       url: 'https://astro.build/',
>     }),
>   };
> }
> ```

Reviewer 2: ``````suggestion
File routes export a `get` function (optionally `async`) which gets called to generate the file. It returns an object with a `body` to save the file contents in your final build. In static mode, Astro will call this at build time to generate the corresponding file.

```js
// Example: src/pages/builtwith.json.ts
// Outputs: /builtwith.json

export async function get({params, request}) {
  return {
    body: JSON.stringify({
      name: 'Astro',
      url: 'https://astro.build/',
    }),
  };
}
```
There's some duplication there to be edited, but I just moved the code comment out into the paragraph itself. I feel like both might have been redundant? Can that literally come out and become part of the lead-in paragraph?

Or was your thinking, "This is for the people who aren't going to read the paragraph, so I'd better put the exact same thing in the code sample?" (I get that! Just does seem to say the same thing twice?)
``````

> The return object can also have an `encoding` property. It can be any valid [`BufferEncoding`](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/bdd02508ddb5eebcf701fdb8ffd6e84eabf47885/types/node/buffer.d.ts#L169) accepted by node.js' `fs.writeFile` method. For example, to produce a binary png image:

Reviewer 2: ```suggestion
The return object can also have an `encoding` property. It can be any valid [`BufferEncoding`](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/bdd02508ddb5eebcf701fdb8ffd6e84eabf47885/types/node/buffer.d.ts#L169) accepted by Node.js' `fs.writeFile` method. For example, to produce a binary png image:
```

> ```ts title="src/pages/astro-logo.png.ts" {7}

Reviewer 3: Small highlighting issue, should be {6}
![CleanShot 2022-09-28 at 18 52 14@2x](https://user-images.githubusercontent.com/19961439/192914162-79dcc5bf-84ba-41bf-a7c0-a1453b245cc9.png)

export async function get({ params, request }) => {
  const response = await fetch("https://astro.build/assets/press/full-logo-light.png");
  const buffer = Buffer.from(await response.arrayBuffer());
  return {
    body: buffer,
    encoding: 'binary',
  };
```

> You can type your endpoint functions using the `APIRoute` type:

Reviewer 2: You can "also"? "instead"?

```js
import type { APIRoute } from 'astro';

export const get: APIRoute = async function get ({params, request}) {
...
```

> As their only argument, endpoint functions recieve an object with two properties: `params` and `request`.

Reviewer 2: I feel like maybe this should go higher, even before all the examples? This isn't just about the `APIRoute` type, right? This is for all of these?

Like up in line 14 maybe:

> File routes export a `get` function (optionally `async`) that receives an object with two properties (`params` and `request`) as its only argument...

### `params` and Dynamic routing

Endpoints support the same [dynamic routing](/en/core-concepts/routing/#dynamic-routes) features that pages do. Name your file with a bracketed parameter name and export a [`getStaticPaths()` function](/en/reference/api-reference/#getstaticpaths). Then, you can access the parameter using the `params` property passed to the endpoint function:

```ts title="src/pages/[id].json.ts"
import type { APIRoute } from 'astro';

> const usernames = ["Sarah", "Chris", "Dan"]

Reviewer 2: 😍

export const get: APIRoute = ({ params, request }) => {
  const id = params.id;
  return {
    body: JSON.stringify({
      name: usernames[id]
    })
  }
};

export function getStaticPaths () {
    return [ 
        { params: { id: "0"} },
        { params: { id: "1"} },
        { params: { id: "2"} },
    ]
}
```

> This will generate three JSON endpoints at build time: `/api/1.json`, `/api/2.json`, `/api/3.json`. Dynamic routing with endpoints works the same as it does with pages, but because the endpoint is a function and not a component, [props](/en/reference/api-reference/#data-passing-with-props) aren't supported.

Reviewer: ```suggestion
This will generate three JSON files at build time: `/api/1.json`, `/api/2.json`, `/api/3.json`. Dynamic routing with endpoints works the same as it does with pages, but because the endpoint is a function and not a component, [props](/en/reference/api-reference/#data-passing-with-props) aren't supported.
```

Reviewer: Also, for SSG mode we probably could support `props`, at least I don't see a reason why not. I don't think we excluded that on purpose.

### `request`

> All endpoints receive a `request` property, but only `request.url` works in static mode. This returns the full URL of the current endpoint, using your [`site`](/en/reference/configuration-reference/#site) config option as the base.

Reviewer: Is this true in dev mode? I would expect that the url is the localhost url during dev.

I would not expect a difference here between endpoints and pages, if there is one it's probably a bug.

> All endpoints receive a `request` property, but only `request.url` works in static mode. This returns the full URL of the current endpoint, using your [`site`](/en/reference/configuration-reference/#site) config option as the base.

Reviewer 2: ```suggestion
All endpoints receive a `request` property, but in static mode, only `request.url` is available. This returns the full URL of the current endpoint, using your [`site`](/en/reference/configuration-reference/#site) config option as the base.
```
I worry about somone scanning and reading "`request.url` only works in static mode" by mistake. So, I've swapped the phrase around to hit BUT IN STATIC MODE (even though I don't *love* the two commas) and I think we can ebe more precise than "works" ... what's correct here, "available"? "have access to"?

```ts title="src/pages/request-path.json.ts"
import type { APIRoute } from 'astro';

export const get: APIRoute = ({ params, request }) => {
  return {
    body: JSON.stringify({
      path: new URL(request.url).pathname
    })
  };
}
```

## Server Endpoints (API Routes)

> Everything in the previous section can be used in SSR mode, but the endpoints will be built when they are requested. This unlocks new features that are unavailable at build time, and allows you to build API routes that listen for requests and securely execute code on the server at runtime.

Reviewer 2: ```suggestion
Everything described in the previous static files section can also be used in SSR mode: files export a `get` function and return a `body`, dynamic routing is available, and all endpoints receive a request property. But, unlike in `static` mode, when you configure `server` mode, the endpoints will be built when they are requested. This unlocks new features that are unavailable at build time, and allows you to build API routes that listen for requests and securely execute code on the server at runtime.
```
Love the second sentence here especially!

The concern I have is that some people are going to jump to this section immediately, not having read the from the top of the page ("I'm doing SSR!"). So, I think this section needs a bit more of an explicit lead in and comparison with what came before. Not totally married to the suggestion above, but something like that I think would make that hit home?

:::note

> Be sure to [enable SSR](http://localhost:3001/en/guides/server-side-rendering/#enabling-ssr-in-your-project) before trying these examples.

Reviewer: this is pointing to localhost

:::

> Endpoints can now access `params` without exporting `getStaticPaths`, and they can return a [`Response`](https://developer.mozilla.org/en-US/docs/Web/API/Response) object, allowing you to set status codes and headers:

Reviewer: ```suggestion
Endpoints can access `params` without exporting `getStaticPaths`, and they can return a [`Response`](https://developer.mozilla.org/en-US/docs/Web/API/Response) object, allowing you to set status codes and headers:
```

:::note
Since `delete` is a reserved word in JavaScript, export a `del` function to match the delete method.
:::

```ts title="src/pages/methods.json.ts"
export const get: APIRoute = ({ params, request }) => {
  return {
    body: JSON.stringify({
      message: "This was a GET!"
    })
  }
};

export const post: APIRoute = ({ request }) => {
  return {
    body: JSON.stringify({
      message: "This was a POST!"
    })
  }
}

export const del: APIRoute = ({ request }) => {
  return {
    body: JSON.stringify({
      message: "This was a DELETE!"
    })
  }
}

export const all: APIRoute = ({ request }) => {
  return {
    body: JSON.stringify({
      message: `This was a ${request.method}!`
    })
  }
}
```

### `request` 
In SSR mode, the `request` property returns a fully usable [`Request`](https://developer.mozilla.org/en-US/docs/Web/API/Request) object that refers to the current request. This allows you to accept data and check headers:

```ts title="src/pages/test-post.json.ts"
export const post: APIRoute = async ({ request }) => {

>   if (request.headers.get("Content-Type") === "application/json") {

Reviewer: good example

    const body = await request.json();
    const name = body.name;
    return new Response(JSON.stringify({
      message: "Your name was: " + name
    }), {
      status: 200
    })
  }
  return new Response(null, { status: 400 });
}
```

### Redirects

> Since `Astro.redirect` is not available in API Routes you can use [`Response.redirect`](https://developer.mozilla.org/en-US/docs/Web/API/Response/redirect) to redirect:

Reviewer: This is a bug and something we'll probably be changing, should we document it anyways since it's useful info for now and just remove later?

Author: Yeah, let's document it for now. Is it just `Astro.redirect` that should be available, or other functions on the Astro global?

> Since `Astro.redirect` is not available in API Routes you can use [`Response.redirect`](https://developer.mozilla.org/en-US/docs/Web/API/Response/redirect) to redirect:

Reviewer 2: ```suggestion
Since `Astro.redirect` is currently not yet available in API Routes you can use [`Response.redirect`](https://developer.mozilla.org/en-US/docs/Web/API/Response/redirect) to redirect:
```
Per Matthew's comment, we could imply that this may yet change, if you wanted.

Author: I think we can just document it as-is, and change it when needed
